graf2d/gpad/src/TPad.cxx

```diff
@@ -3067,30 +3067,73 @@ Bool_t TPad::Collide(Int_t i, Int_t j, Int_t w, Int_t h)
 ///
 /// \return `true` if the box could be placed, `false` if not.
 ///
-/// \param[in]  o        pointer to the box to be placed
-/// \param[in]  w        box width to be placed
-/// \param[in]  h        box height to be placed
-/// \param[out] xl       x position of the bottom left corner of the placed box
-/// \param[out] yb       y position of the bottom left corner of the placed box
+/// \param[in]  o               pointer to the box to be placed
+/// \param[in]  w               box width to be placed
+/// \param[in]  h               box height to be placed
+/// \param[out] xl              x position of the bottom left corner of the placed box
+/// \param[out] yb              y position of the bottom left corner of the placed box
+/// \param[in]  options         l=left, r=right, t=top, b=bottom, w=within margins. Order determines
+///                             priority for placement. Default is "lb" (prioritises horizontal over vertical)
 
-Bool_t TPad::PlaceBox(TObject *o, Double_t w, Double_t h, Double_t &xl, Double_t &yb)
+Bool_t TPad::PlaceBox(TObject *o, Double_t w, Double_t h, Double_t &xl, Double_t &yb, Option_t* opt)
 {
    FillCollideGrid(o);
 
    Int_t iw = (int)(fCGnx*w);
    Int_t ih = (int)(fCGny*h);
 
+   Int_t nxmin = 0;
+   Int_t nymin = 0;
    Int_t nxmax = fCGnx-iw-1;
    Int_t nymax = fCGny-ih-1;
+   Int_t dx = 1;
+   Int_t dy = 1;
 
-   for (Int_t i = 0; i<nxmax; i++) {
-      for (Int_t j = 0; j<=nymax; j++) {
-         if (Collide(i,j,iw,ih)) {
-            continue;
-         } else {
-            xl = (Double_t)(i)/(Double_t)(fCGnx);
-            yb = (Double_t)(j)/(Double_t)(fCGny);
-            return kTRUE;
+   TString sOpt(opt);
+   sOpt.ToLower();
+
+   if (sOpt.Contains("w")) {
+      nxmin += fCGnx*GetLeftMargin();
+      nymin += fCGny*GetBottomMargin();
+      nxmax -= fCGnx*GetRightMargin();
+      nymax -= fCGny*GetTopMargin();
+   }
+
+   if (sOpt.Contains("t")) {
+      // go from top to bottom instead of bottom to top
+      dy = -1;
+   }
+
+   if (sOpt.Contains("r")) {
+      // go from right to left instead of left to right
+      dx = -1;
+   }
+
+   if ( std::max(sOpt.Index("t"),sOpt.Index("b")) < std::max(sOpt.Index("l"),sOpt.Index("r")) ) {
```

Comment:
I don't particularly like this, as the repeated calls to `TString::Index()` traverse the string many times.  Maybe we could do it in a single `for` loop above -- also I think the resulting code is more readable.  :slightly_smiling_face:

Reply:
Well, I didn't think it was that bad, but would you be happier with something like:

```
bool prioritizeVertical = false;
for(size_t i=0; i < sOpt.Length(); i++) {
  if(sOpt[i] == 't' || sOpt[i] == 'b') {
    prioritizeVertical = true; break;
  } else if(sOpt[i] == 'l' || sOpt[i] == 'r') {
    break;
  }
}
```

Reply:
Maybe even a single loop that also catches the behavior of the `if` statements in lines https://github.com/root-project/root/pull/12066/files#diff-2651ca80f64ee138111656abb5bd6e33498cfb7081f8a31df5f2c77052ce004bR3095-R3110 (?), so that the list of options is traversed only once.
Thanks again for the contribution!

Reply:
a single loop adding those bits would be hard because, as you can see, i am breaking out of the loop in order to determine the prioritization. If you still have strong feelings about how this should be done perhaps can you put it in the form of a code suggestion I could then just accept if its functionally equivalent. Is this ok?

Reply:
How should we proceed with this one? Conscious that 6.28 is only a few days away from being released now ...

Reply:
I am discussing a solution with @guitargeek; we plan to merge these changes today/tomorrow.